### PR TITLE
Adding the Cauchy distribution to torch.distributions

### PR DESCRIPTION
**Pull request by neerajprad**
This adds the Cauchy distribution to `torch.distributions`. Note that `sample` uses the default implementation since the implementation of `cauchy` (unlike `normal`) in PyTorch does not allow for mu/gamma parameters to be tensors.

### CODE REVIEW

File: torch/distributions/cauchy.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import torch
+from torch.distributions import Distribution
+from torch.distributions.utils import broadcast_all
+
+
+class Cauchy(Distribution):
+    r"""
+    Samples from a Cauchy (Lorentz) distribution. The distribution of the ratio of
+    independent normally distributed random variables with means `0` follows a
```
Comment (fritzo) [inline torch/distributions/cauchy.py]: Is this "follows" or "roughly follows"? IIRC this distribution was approximate.

File: torch/distributions/cauchy.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    Args:
+        loc (float or Tensor or Variable): mode or median of the distribution.
+        scale (float or Tensor or Variable): half width at half maximum.
+    """
+    has_rsample = True
+
+    def __init__(self, loc, scale):
+        self.loc, self.scale = broadcast_all(loc, scale)
+        if isinstance(loc, Number) and isinstance(loc, Number):
```
Comment (fritzo) [inline torch/distributions/cauchy.py]: typo, replace second 'loc' with 'scale'

Comment (neerajprad) [reply torch/distributions/cauchy.py]: done

File: torch/distributions/cauchy.py (hunk 1 above)
Comment (neerajprad) [reply torch/distributions/cauchy.py]: I thought it wasn't an approximation; there don't seem to be any assumptions -
 http://mathworld.wolfram.com/NormalRatioDistribution.html, but let me know if I missed something.

File: test/test_distributions.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 from torch.autograd import Variable, gradcheck
 from torch.distributions import (Bernoulli, Beta, Categorical, Dirichlet,
                                  Exponential, Gamma, Laplace, Normal)
+from torch.distributions.cauchy import Cauchy
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply test/test_distributions.py]: My bad, I missed the documentation.

File: torch/distributions/cauchy.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/cauchy.py]: Oh nice, you're right. I was just confused by the wording of http://docs.pyro.ai/primitive_dist.html#module-pyro.distributions.cauchy

File: test/test_distributions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fritzo) [reply test/test_distributions.py]: LGTM as soon as this comment is addressed

File: torch/distributions/cauchy.py (hunk 1 above)
Comment (neerajprad) [reply torch/distributions/cauchy.py]: Yeah, I am not sure about that either. :)  We should change it too, or deprecate it soon.

**Comment by neerajprad** [comment]
Thanks for the review @fritzo. Sent to pytorch for review.

**Pull request closed without merging**